extensions/amp-form/0.1/form-dirtiness.js

```diff
@@ -0,0 +1,168 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {isDisabled, isFieldDefault, isFieldEmpty} from '../../../src/form';
+
+export const DIRTINESS_INDICATOR_CLASS = 'amp-form-dirty';
+
+export class FormDirtiness {
+  /**
+   * @param {!HTMLFormElement} form
+   */
+  constructor(form) {
+    /** @private @const {!HTMLFormElement} */
+    this.form_ = form;
+
+    /** @private {number} */
+    this.dirtyFieldCount_ = 0;
+
+    /** @private {!Object<string, boolean>} */
+    this.isFieldNameDirty_ = {};
+
+    /** @private {boolean} */
+    this.isSubmitting_ = false;
+
+    this.installEventListeners_();
+  }
+
+  /**
+   * Processes dirtiness state when a form is being submitted.
+   */
+  onSubmitting() {
+    this.isSubmitting_ = true;
+    this.updateDirtinessClass_();
+  }
+
+  /**
+   * Processes dirtiness state when the form submission fails.
+   */
+  onSubmitError() {
+    this.isSubmitting_ = false;
+    this.updateDirtinessClass_();
+  }
+
+  /**
+   * Processes dirtiness state when the form submission succeeds.
+   */
+  onSubmitSuccess() {
+    this.isSubmitting_ = false;
+    this.clearDirtyFields_();
+    this.updateDirtinessClass_();
+  }
+
+  /**
+   * @private
+   */
+  installEventListeners_() {
+    this.form_.addEventListener('input', this.onInput_.bind(this));
+    this.form_.addEventListener('reset', this.onReset_.bind(this));
+  }
+
+  /**
+   * @param {!Event} event
+   * @private
+   */
+  onInput_(event) {
+    const field = dev().assertElement(event.target);
+    this.checkDirtinessAfterUserInteraction_(field);
+    this.updateDirtinessClass_();
+  }
+
+  /**
+   * @param {!Event} unusedEvent
+   * @private
+   */
+  onReset_(unusedEvent) {
+    this.clearDirtyFields_();
+    this.updateDirtinessClass_();
+  }
+
+  /**
+   * @param {!Element} field
+   * @private
+   */
+  checkDirtinessAfterUserInteraction_(field) {
+    if (shouldSkipDirtinessCheck(field)) {
+      return;
+    }
+
+    if (isFieldEmpty(field) || isFieldDefault(field)) {
+      this.removeDirtyField_(field.name);
+    } else {
+      this.addDirtyField_(field.name);
+    }
+  }
+
+  /**
+   * @param {string} fieldName
+   * @private
+   */
+  addDirtyField_(fieldName) {
+    if (!this.isFieldNameDirty_[fieldName]) {
+      this.isFieldNameDirty_[fieldName] = true;
+      ++this.dirtyFieldCount_;
+    }
+  }
+
+  /**
+   * @param {string} fieldName
+   * @private
+   */
+  removeDirtyField_(fieldName) {
+    if (this.isFieldNameDirty_[fieldName]) {
+      delete this.isFieldNameDirty_[fieldName];
+      --this.dirtyFieldCount_;
+    }
+  }
+
+  /**
+   * @private
+   */
+  clearDirtyFields_() {
+    this.isFieldNameDirty_ = {};
+    this.dirtyFieldCount_ = 0;
+  }
+
+  /**
+   * @private
+   */
+  updateDirtinessClass_() {
+    if (this.dirtyFieldCount_ == 0 || this.isSubmitting_) {
```

Comment:
This can be simplified into 

```js
const isDirty = this.dirtyFieldCount_ > 0 && !this.isSubmitting_;
this.form_.classList.toggle(DIRTINESS_INDICATOR_CLASS, isDirty);
```